docs/okeps/okep-5672-metadata-service.md

```diff
@@ -0,0 +1,245 @@
+# OKEP-5672: Metadata Service
+
+* Issue: [#5672](https://github.com/ovn-org/ovn-kubernetes/issues/5672)
+
+## Problem Statement
+
+VM cloud platforms typically offer metadata service at 169.254.169.254.
+For Kubevirt VMs running on ovn-kubernetes, we would like to offer the same kind of support.
+
+## Goals
+
+* Allow a user to configure metadata service access with their UDN that enables access to 169.254.169.254 with Internal Traffic Policy (ITP) local
+* Limit access to this IP within the SDN (ovn-kubernetes networked pods)
+
+## Non-Goals
+
+* Standardizing any kind of REST API, or metadata pod backend for the service
+* IPv6 Support
+
+## Introduction
+
+Most VM cloud platform providers expose a metadata service to VMs. This service is used by the VM to get cloud-init
+data at boot time, as well as to query for metadata post-boot. The providers all consistently use 169.254.169.254 as
+the IP to access metadata. However, each platform uses different REST API to query this metadata. Kubevirt offers
+cloud-init via a [NoCloud](https://cloudinit.readthedocs.io/en/latest/reference/datasources/nocloud.html) option, but
+this does not provide a dynamic source that can be queried post-boot.
+
+The goal of this feature is to provide a way that a user can specify to OVN-Kubernetes that they would like to use
+a metadata service provided at 169.254.169.254. OVN-Kubernetes will handle creating an OVN Load Balancer that will use
+169.254.169.254 as the VIP, and then the user can specify backends. How these backends are implemented is beyond the
+scope of this enhancement or OVN-Kubernetes. In the future it may be standardized by a higher layer, like Kubevirt.
+
+Note there is limited support for metadata service on VM cloud platforms for IPv6. There is also no consistent IPv6
+address used. For that reason, IPv6 support will be deferred from this enhancement.
+
+## User-Stories/Use-Cases
+
+Story 1: Provide Dynamic Metadata to a VM
+
+As a tenant using Kubevirt, I want to enable my VMs to be able to query a metadata service I manage as pods
+so that my VM can get updated token information during runtime. The VM should be able to query 169.254.169.254
+to get dynamically updated information.
+
+## Proposed Solution
+
+This proposal includes specifying new API within the UDN CRD spec to allow a user to specify the metadata service
+name in their UDN. Once identified, OVN-Kubernetes will add an OVN Load Balancer with the 169.254.169.254 VIP for
+this service.
```

Comment:
When we chatted on Slack, I thought we were proposing using https://ovn-kubernetes.io/okeps/okep-5233-preconfigured-udn-addresses/ and extending those concepts from L2 UDNs to apply to L3 UDNs too, given we think that it could be required for EVPN.

So the UX would be:

1. Add `reservedSubnets: ["169.254.169.254/32"]` to your UDN
2. When deploying your metadata service pod add this annotation: `v1.multus-cni.io/default-network: |
      {"name": "default", "namespace": "ovn-kubernetes", "ips": ["169.254.169.254"], "mac": "00:1A:2B:3C:4D:5E", "ipam-claim-reference": "my-claim"}`

Reply:
Yeah then I realized it wont work cause we need a pod per node attached to the UDN, and they cant all have the same IP, especially on a layer 2 network. So i went back to the service approach, but made it a singleton in the API so I thought it would be OK.